aggregators/codec.go

```diff
@@ -82,14 +83,14 @@ func (k *CombinedMetricsKey) SizeBinary() int {
 // ToProto converts CombinedMetrics to its protobuf representation.
 func (m *combinedMetrics) ToProto() *aggregationpb.CombinedMetrics {
 	pb := aggregationpb.CombinedMetricsFromVTPool()
-	if len(m.Services) > cap(pb.ServiceMetrics) {
-		pb.ServiceMetrics = make([]*aggregationpb.KeyedServiceMetrics, 0, len(m.Services))
-	}
+	pb.ServiceMetrics = slices.Grow(pb.ServiceMetrics, len(m.Services))
 	for k, m := range m.Services {
-		ksm := aggregationpb.KeyedServiceMetricsFromVTPool()
-		ksm.Key = k.ToProto()
-		ksm.Metrics = m.ToProto()
-		pb.ServiceMetrics = append(pb.ServiceMetrics, ksm)
+		pb.ServiceMetrics = pb.ServiceMetrics[:len(pb.ServiceMetrics)+1]
+		if pb.ServiceMetrics[len(pb.ServiceMetrics)-1] == nil {
+			pb.ServiceMetrics[len(pb.ServiceMetrics)-1] = &aggregationpb.KeyedServiceMetrics{}
+		}
+		pb.ServiceMetrics[len(pb.ServiceMetrics)-1].Key = k.ToProto()
+		pb.ServiceMetrics[len(pb.ServiceMetrics)-1].Metrics = m.ToProto()
```

Comment:
Why do we need to do this? Why not just directly append?

Reply:
IIANM to get the most benefit we should reuse the existing slice elements. So instead of assigning the result of ToProto, we would need to update the target pb type fields from the intermediate type.

I'll have a stab at that while Carson is sleeping.

Reply:
@lahsivjar to answer your question though, the change is to reuse existing space in the slices:
- the Grow call ensures there's capacity (which may be existing/reused space)
- the reslice on line 88 ensures the slice _length_ is extended into the capacity
- then we index that element directly

The change I mentioned above would be a further optimisation to reuse the existing entry, if non-nil, instead of getting a new value from the pool

Reply:
Ah, got it. I think it might be better to just reslice once and then use the array semantics for a more readable code.

Something like:

```go
	pb.ServiceMetrics = pb.ServiceMetrics[:len(m.Services)]
	var i int
	for k, m := range m.Services {
		if pb.ServiceMetrics[i] == nil {
			pb.ServiceMetrics[i] = &aggregationpb.KeyedServiceMetrics{}
		}
		pb.ServiceMetrics[i].Key = k.ToProto()
		pb.ServiceMetrics[i].Metrics = m.ToProto()
		i++
	}
```

Reply:
> IIANM to get the most benefit we should reuse the existing slice elements. So instead of assigning the result of ToProto, we would need to update the target pb type fields from the intermediate type.

This doesn't work: although the elements are not set to nil, they are reset. Resetting sends the key and value back to their pools.

Reply:
That should be because `KeyedServiceMetrics` qualifies as a `ShouldPool` method. If we were to make `KeyedServiceMetrics` not pool we will be able to do that.

Reply:
To do the above we have 2 choices:

1. Remove `KeyedServiceMetrics` explicitly from the list while generation
2. Mark each message that require pool explicitly in the proto definition

`2` can be done by:

```proto
message {
  option (vtproto.mempool) = true;
  ...
}
```

Reply:
Above being said, I am not sure if it would be a good idea. A few benchmarks should help make the decisions but without those I think just backing the `.Key` and `.Metrics` with pool (i.e. what is done right now) should be okay.

Reply:
I'll abandon this for now, I think it's not likely to yield much. Will get back to looking into parallelism.

@carsonip I think it would be good to tidy up the reslicing and indexing, like @lahsivjar suggested in https://github.com/elastic/apm-aggregation/pull/63#discussion_r1281365678. Then we can come back to the rest later, maybe.

Reply:
Thanks @lahsivjar  for the good suggestion.

This rather unappealing piece (at this for this occasion) of code came from vtproto and it was written for Unmarshal in this way since it had no certainly for the length of new slice. In our code, we do, and I added a slices.Grow to ensure there was only 1 alloc if `len(newSlice) < cap(reusedSlice)`. With this `slices.Grow` call, we can do what @lahsivjar suggested, to `pb.ServiceMetrics = pb.ServiceMetrics[:len(m.Services)]` once and remove that reslicing and indexing code.

Reply:
On a side note, even though `KeyedServiceMetrics` qualifies as a `ShouldPool`, vtproto Marshal / Unmarshal ignores that and does not use the pool. I pooled individual elements in [my fork](https://github.com/carsonip/vtprotobuf/tree/fix-pooling-unmarshal-from-pool), but it seems that not pooling individual elements is faster. The alternative is implemented in #37